h2o-core/src/main/java/water/api/RemoveAllHandler.java

```diff
@@ -23,14 +20,47 @@ public RemoveAllV3 remove(int version, RemoveAllV3 u) {
       RapidsHandler.SESSIONS.clear();
     }
     fs.blockForPending();
+
+    if (u.retained_keys != null && u.retained_keys.length != 0) {
+      retainKeys(u.retained_keys);
+    } else {
+      clearAll();
+    }
+    Log.info("Finished removing objects");
+    return u;
+  }
+
+
+  private void clearAll() {
+    Log.info("Removing all objects");
     // Bulk brainless key removal.  Completely wipes all Keys without regard.
-    new MRTask(H2O.MIN_HI_PRIORITY){
-      @Override public void setupLocal() {  H2O.raw_clear();  water.fvec.Vec.ESPC.clear(); }
+    new MRTask(H2O.MIN_HI_PRIORITY) {
+      @Override
+      public void setupLocal() {
+        H2O.raw_clear();
+        water.fvec.Vec.ESPC.clear();
+      }
     }.doAllNodes();
     // Wipe the backing store without regard as well
     H2O.getPM().getIce().cleanUp();
     H2O.updateNotIdle();
-    Log.info("Finished removing objects");
-    return u;
   }
-}
\ No newline at end of file
+
+  private void retainKeys(final KeyV3[] retained_keys) {
+    Log.info(String.format("Removing all objects, except for %d provided key(s)", retained_keys.length));
```

Comment:
possible NPE? line 52 permits `retained_keys` to be null

Reply:
This is actually secured on line 24 `if (u.retained_keys != null && u.retained_keys.length != 0) {`.

Reply:
ok, I see - I got confused by the naming `retained_keys` vs `retainedKeys`

makes sense, sorry